federation/planner.go

```diff
@@ -0,0 +1,300 @@
+package federation
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+
+	"github.com/samsarahq/thunder/graphql"
+)
+
+type StepKind int
+
+const (
+	KindType StepKind = iota
+	KindField
+)
+
+type PathStep struct {
+	Kind StepKind
+	Name string
+}
+
+type Plan struct {
+	Path    []PathStep
+	Service string
+	Kind    string
+	// XXX: What are we using Type for here again? -- oh, it's for the __federation field...
+	Type         string
+	SelectionSet *graphql.SelectionSet
+	After        []*Plan
+}
+
+type Executor struct {
+	schema    *SchemaWithFederationInfo
+	flattener *flattener
+}
+
+
+func (e *Executor) planObject(typ *graphql.Object, selectionSet *graphql.SelectionSet, service string) (*Plan, error) {
+	p := &Plan{
+		Type:         typ.Name,
+		Service:      service,
+		SelectionSet: &graphql.SelectionSet{},
+		After:        nil,
+		Kind:         "query",
+	}
+
+	// XXX: pass in sub-path (and sub-plan slice?) to make sub-plan munging simpler?
+	// - should we type check here?
+
+	// - refactor to return array of subplans with preferential treatment for given service?
+	// eh whatever.
+
+	var localSelections []*graphql.Selection
+	selectionsByService := make(map[string][]*graphql.Selection)
+
+	// Assert that we are working with a flattened query.
+	if len(selectionSet.Fragments) > 0 {
+		return nil, errors.New("selectionSet has fragments, expected flattened query")
+	}
+
+	for _, selection := range selectionSet.Selections {
+		if selection.Name == "__typename" {
```

Comment:
I'm not sure I understand how "__typename" is being used (what does it mean, who sets it, etc.).  Maybe this would be a good thing to document somewhere?  (top-level comment in this file?)